### PR TITLE
DEP: GeoSeries.geom_almost_equals

### LINKED ISSUE #2538: Deprecate GeoSeries.almost_equals (deprecated in shapely)
The `almost_equals` method in Shapely is deprecated (and will be removed in a future version): https://github.com/shapely/shapely/blob/6c896a7fb0348f46630e4045e1fdc3e54212dc42/shapely/geometry/base.py#L666-L669

So we should probably follow suite in geopandas as well and deprecate our almost_equals method in favor of its alias `equals_exact`

### PR DESCRIPTION
**Pull request by darshanip**
Fixes #2538

Need confirmation, apart from the deletion I did do we need to remove 'geom_almost_equals' and its usages as well? 
This is first **code** contribution in OS.

### CODE REVIEW

**Comment by m-richards** [comment]
Hi @darshanip, this is unused code so is good to delete. With geom_almost_equals, we won't remove that yet, but should rather add a FutureWarning to indicate to users it will be removed in a future version.

The warnings removed in #2267 are an example of the kind of message you should add here if you'd like some guidance.

**Comment by darshanip** [comment]
@m-richards Please review the changes.

**Comment by darshanip** [comment]
@jorisvandenbossche @martinfleis, Could you please review changes.

File: geopandas/array.py [hunk 1]
```diff
@@ -566,6 +566,12 @@ def geom_equals_exact(self, other, tolerance):
         return self._binary_method("equals_exact", self, other, tolerance=tolerance)
 
     def geom_almost_equals(self, other, decimal):
+        warnings.warn(
+            "geom_almost_equals() is now geom_equals_exact(). "
+            "geom_almost_equals() will be deprecated in the future.",
```
Comment (m-richards) [inline geopandas/array.py:571]: I think it would make sense to be more explicit here (borrowing from the underlying shapely warning)
> "The 'geom_almost_equals()' method is deprecated because the name is
            confusing. The 'equals_exact()' method should be used
            instead."

File: geopandas/base.py [hunk 2]
```diff
@@ -1057,6 +1058,12 @@ def geom_almost_equals(self, other, decimal=6, align=True):
         GeoSeries.geom_equals_exact
 
         """
+        warnings.warn(
+            "geom_almost_equals() is now geom_equals_exact(). "
```
Comment (m-richards) [inline geopandas/base.py:1062]: I would change this as well as above.

**Review by m-richards — COMMENTED** [review]
Hi @darshanip, sorry for the delayed response.

This is looking good so far but there is a little bit more to do:
* We should update the tests using `geom_almost_equals` to use `geom_equals_exact` instead and the helper method in `testing.py`
* In `test_geoseries::test_geom_almost_equals` we should add a test to check the correct warning is emitted, and filter the warning from the other usages.
* We should update the documentation in `data_structures.rst` to refer to geom_equals_exact instead.

File: geopandas/base.py (hunk 2 above)
Comment (darshanip) [reply geopandas/base.py]: done.

File: geopandas/array.py (hunk 1 above)
Comment (darshanip) [reply geopandas/array.py]: done.

**Comment by darshanip** [comment]
@m-richards can you please review code.

File: geopandas/tests/test_geoseries.py [hunk 3]
```diff
@@ -148,14 +148,14 @@ def test_geom_equals_align(self):
...
+        assert_array_equal(self.g1.geom_equals_exact(self.sq), [False, True])
 
         assert_array_equal(
-            self.a1.geom_almost_equals(self.a2, align=True), [False, True, False]
+            self.a1.geom_equals_exact(self.a2, align=True), [False, True, False]
         )
         assert_array_equal(
-            self.a1.geom_almost_equals(self.a2, align=False), [False, False]
+            self.a1.geom_equals_exact(self.a2, align=False), [False, False]
         )
```
Comment (martinfleis) [inline geopandas/tests/test_geoseries.py:159]: Can you revert this change? The method is deprecated but not gone yet, so we still should test if it works as expected. The best would be to wrap the original code in

```py
with pytest.warns(FutureWarning, match="The 'geom_almost_equals()'"):
    ...
```

File: doc/source/docs/user_guide/data_structures.rst [hunk 4]
```diff
@@ -72,7 +72,7 @@ Basic methods
 Relationship tests
 ^^^^^^^^^^^^^^^^^^^
 
-* :meth:`~GeoSeries.geom_almost_equals`: is shape almost the same as ``other`` (good when floating point precision issues make shapes slightly different)
+* :meth:`~GeoSeries.geom_equals_exact`: is shape almost the same as ``other`` (good when floating point precision issues make shapes slightly different)
```
Comment (m-richards) [inline doc/source/docs/user_guide/data_structures.rst:75]:
```suggestion
* :meth:`~GeoSeries.geom_equals_exact`: is shape the same as ``other`` (up to a specified decimal place tolerance)
```

File: geopandas/base.py [hunk 5]
```diff
@@ -1140,13 +1140,13 @@ def geom_almost_equals(self, other, decimal=6, align=True):
...
-        >>> s.geom_almost_equals(Point(0, 1), decimal=2)
+        >>> s.geom_equals_exact(Point(0, 1), decimal=2)
         0    False
         1    False
         2     True
         dtype: bool
 
-        >>> s.geom_almost_equals(Point(0, 1), decimal=1)
+        >>> s.geom_equals_exact(Point(0, 1), decimal=1)
         0    False
```
Comment (m-richards) [inline geopandas/base.py:1150]: These need to be reverted, they are the documentation for `geom_almost_equals`, which still needs to be accurate until `geom_almost_equals` is completely removed

File: geopandas/testing.py [hunk 6]
```diff
@@ -90,7 +90,7 @@ def _geom_almost_equals_mask(this, that):
     """
 
     return (
-        this.geom_almost_equals(that)
+        this.geom_equals_exact(that)
```
Comment (m-richards) [inline geopandas/testing.py:93]: This doesn't work as is because `geom_equals_exact` requires tolerance be specified. For this use we should directly translate the equivalent of the old `geom_almost_equals`:
```suggestion
        this.geom_equals_exact(that, tolerance=0.5 * 10 ** (-6))
```

File: geopandas/testing.py [hunk 7]
```diff
@@ -144,7 +144,7 @@ def assert_geoseries_equal(
         Check that both are same type (*and* are GeoSeries). If False,
         will attempt to convert both into GeoSeries.
     check_less_precise : bool, default False
-        If True, use geom_almost_equals. if False, use geom_equals.
+        If True, use geom_equals_exact. if False, use geom_equals.
```
Comment (m-richards) [inline geopandas/testing.py:147]:
```suggestion
        If True, use geom_equals_exact. if False, use geom_equals.
```
```suggestion
        If True, use geom_equals_exact with relative error of 0.5e-6. if False, use geom_equals.
```

File: geopandas/tests/test_dissolve.py [hunk 8]
```diff
@@ -61,7 +61,7 @@ def expected_mean(merged_shapes):
 def test_geom_dissolve(nybb_polydf, first):
     test = nybb_polydf.dissolve("manhattan_bronx")
     assert test.geometry.name == "myshapes"
-    assert test.geom_almost_equals(first).all()
+    assert test.geom_equals_exact(first).all()
```
Comment (m-richards) [inline geopandas/tests/test_dissolve.py:64]: This also needs to be updated to provide the second parameter for `geom_equals_exact`

File: geopandas/tests/test_geom_methods.py [hunk 9]
```diff
@@ -870,10 +870,10 @@ def test_rotate(self):
 
         o = Point(0, 0)
         res = self.g4.rotate(angle, origin=o).rotate(-angle, origin=o)
-        assert geom_almost_equals(self.g4, res)
+        assert geom_equals_exact(self.g4, res)
```
Comment (m-richards) [inline geopandas/tests/test_geom_methods.py:873]: These are also all missing arguments. I would suggest updating the implementation of `geopandas.tests.util.geom_almost_equals` to use `geom_equals_exact` and then renaming the function to `_geom_almost_equals` to indicate it is private. Then all the usages in `test_geom_methods` can be otherwise unchanged.

File: geopandas/tests/test_geoseries.py [hunk 10]
```diff
@@ -202,7 +203,7 @@ def test_representative_point(self):
     def test_transform(self):
         utm18n = self.landmarks.to_crs(epsg=26918)
         lonlat = utm18n.to_crs(epsg=4326)
-        assert np.all(self.landmarks.geom_almost_equals(lonlat))
+        assert np.all(self.landmarks.geom_equals_exact(lonlat))
```
Comment (m-richards) [inline geopandas/tests/test_geoseries.py:206]: all missing tolerance arg / migrate to use geom_almost_equals from geopandas.tests.util

File: geopandas/base.py [hunk 11]
```diff
@@ -1163,6 +1163,12 @@ def geom_almost_equals(self, other, decimal=6, align=True):
         GeoSeries.geom_equals_exact
 
         """
+        warnings.warn(
+            "The 'geom_almost_equals()' method is deprecated because the name is"
+            "confusing. The 'equals_exact()' method should be used instead.",
```
Comment (m-richards) [inline geopandas/base.py:1168]:
```suggestion
            "confusing. The 'geom_equals_exact()' method should be used instead.",
```

File: geopandas/base.py [hunk 12]
```diff
@@ -1228,7 +1234,6 @@ def geom_equals_exact(self, other, tolerance, align=True):
         See also
         --------
         GeoSeries.geom_equals
-        GeoSeries.geom_almost_equals
```
Comment (m-richards) [inline geopandas/base.py:1231]: It would be good to also update the docstring for the tolerance parameter as it is currently incorrect, it is a relative error tolerance, rather decimal place precision.

**Review by m-richards — COMMENTED** [review]
Hi @darshanip, I've left a handful of further comments which should help with fixing some of the test errors.

File: geopandas/testing.py [hunk 13]
```diff
@@ -144,7 +144,7 @@ def assert_geoseries_equal(
         Check that both are same type (*and* are GeoSeries). If False,
         will attempt to convert both into GeoSeries.
     check_less_precise : bool, default False
-        If True, use geom_equals_exact. if False, use geom_equals.
+        If True, use geom_equals_exact with relative error of 0.5e-6. if False, use geom_equals.
```
Comment (martinfleis) [inline geopandas/testing.py:147]:
```suggestion
        If True, use geom_equals_exact with relative error of 0.5e-6.
        If False, use geom_equals.
```
This shall fix the linting issue.

**Comment by m-richards** [comment]
I've revived this, should be close to ready now. I'm not sure if we should also deprecate geopandas.testing.geom_almost_equals. I don't really see a huge issue with keeping it. If we were to deprecate, it would definitely be worth keeping a private internal helper for this.

File: CHANGELOG.md [hunk 14]
```diff
@@ -23,6 +23,10 @@ Bug fixes:
 - Fix bug in `apply` with `axis=1` where the given user defined function returns nested 
   data in the geometry column (#2959)
 
+Deprecations and compatibility notes:
+ - ``GeometryArray.geom_almost_equals()`` has been deprecated and
```
Comment (martinfleis) [inline CHANGELOG.md:27]:
```suggestion
 - ``geom_almost_equals()`` methods have been deprecated and  
```
This would be probably more precise as it is deprecated on both GeometryArray and GeoPandasBase.

**Review by martinfleis — APPROVED** [review]
Thanks!

> I'm not sure if we should also deprecate geopandas.testing.geom_almost_equals.

Fine with keeping it as well.

**Comment by m-richards** [comment]
> We should have it one one level only, probably on array. We could also keep it on both level but use `geom_equals_exact` array-level method in base-level `geom_almost_equals`.

I've done the latter, it seems better to keep the warning in base.py so the stacklevel is correct for an end user.

**Comment by m-richards** [comment]
@martinfleis I've had to continue this at #3017 as I seem to no longer have push permissions on the fork.

**Comment by martinfleis** [comment]
Superseded by https://github.com/geopandas/geopandas/pull/3017
